lib/full-site-editing/block-templates.php

```diff
@@ -0,0 +1,333 @@
+}
+
+/**
+ * Build a unified template object based on a theme file.
+ *
+ * @param array $template_file Theme file.
+ * @param array $template_type wp_template or wp_template_part.
+ *
+ * @return WP_Block_Template Template.
+ */
+function _gutenberg_build_template_result_from_file( $template_file, $template_type ) {
+	$default_template_types = gutenberg_get_default_template_types();
+
+	$theme               = wp_get_theme()->get_stylesheet();
+	$template            = new WP_Block_Template();
+	$template->id        = $theme . '//' . $template_file['slug'];
+	$template->theme     = $theme;
+	$template->content   = file_get_contents( $template_file['path'] );
+	$template->slug      = $template_file['slug'];
+	$template->is_custom = false;
+	$template->type      = $template_type;
+	$template->title     = $template_file['slug'];
+	$template->status    = 'publish';
+
+	if ( 'wp_template' === $template_type && isset( $default_template_types[ $template_file['slug'] ] ) ) {
+		$template->description = $default_template_types[ $template_file['slug'] ]['description'];
+		$template->title       = $default_template_types[ $template_file['slug'] ]['title'];
+	}
+
+	return $template;
+}
+
+/**
+ * Build a unified template object based a post Object.
+ *
+ * @param WP_Post $post Template post.
+ *
+ * @return WP_Block_Template|WP_Error Template.
+ */
+function _gutenberg_build_template_result_from_post( $post ) {
+	$terms = get_the_terms( $post, 'wp_theme' );
+
+	if ( is_wp_error( $terms ) ) {
+		return $terms;
+	}
+
+	if ( ! $terms ) {
+		return new WP_Error( 'template_missing_theme', __( 'No theme is defined for this template.', 'gutenberg' ) );
+	}
+
+	$theme = $terms[0]->name;
+
+	$template              = new WP_Block_Template();
+	$template->wp_id       = $post->ID;
+	$template->id          = $theme . '|' . $post->post_name;
+	$template->theme       = $theme;
+	$template->content     = $post->post_content;
+	$template->slug        = $post->post_name;
+	$template->is_custom   = true;
+	$template->type        = $post->post_type;
+	$template->description = $post->post_excerpt;
+	$template->title       = $post->post_title;
+	$template->status      = $post->post_status;
+
+	return $template;
+}
+
+/**
+ * Retrieves a list of unified template objects based on a query.
+ *
+ * @param array $query {
+ *     Optional. Arguments to retrieve templates.
+ *
+ *     @type array  $slug__in List of slugs to include.
+ *     @type int    $wp_id Post ID of customized template.
+ * }
+ * @param array $template_type wp_template or wp_template_part.
+ *
+ * @return array Templates.
+ */
+function gutenberg_get_block_templates( $query = array(), $template_type = 'wp_template' ) {
+	$wp_query_args = array(
+		'post_status'    => array( 'auto-draft', 'draft', 'publish' ),
+		'post_type'      => $template_type,
+		'posts_per_page' => -1,
+		'no_found_rows'  => true,
+		array(
+			'taxonomy' => 'wp_theme',
+			'field'    => 'name',
+			'terms'    => $query['theme'],
+		),
```

Comment:
This is incorrect, and wouldn't query by taxonomy.

It's noticeable if you try saving a custom front-page on a theme, then switch to another theme: in the editor, it tries to resolve the custom template from the first theme; in the front end it just simply fails.

```suggestion
		'tax_query' => array(
			array(
				'taxonomy' => 'wp_theme',
				'field'    => 'name',
				'terms'    => $query['theme'],
			),
		),
```

Reply:
Oh Right 👍

Reply:
For some reason, fixing the query is breaking the php unit tests, it seems when `tax_input` is used in php unit tests, the terms are not actually saved properly for some reason.

Anyone has an idea here. cc @TimothyBJacobs

Reply:
Fixed :) There were some logic issues.